### PR TITLE
Floating Button

### PR DESCRIPTION
**Pull request by Matias-Villalobo**
In this commit i have implemented a floating button that load the data store in the local data base

https://user-images.githubusercontent.com/73836200/111829048-f62f5300-88ca-11eb-9582-d9b3229c0180.mp4

### CODE REVIEW

File: app/src/main/res/layout/activity_main.xml [hunk 1]
```diff
@@ -17,6 +17,18 @@
         android:layout_width="match_parent"
         app:layout_constraintBottom_toBottomOf="parent"
         app:layout_constraintTop_toTopOf="parent" />
+
+    <com.google.android.material.floatingactionbutton.FloatingActionButton
+        android:id="@+id/fab"
+        android:layout_width="@dimen/fab_width"
+        android:layout_height="wrap_content"
```
Comment (mauroLonghi) [inline app/src/main/res/layout/activity_main.xml]: Pls use the same values for width and height : 60x60 or wrapxwrap

File: app/src/main/java/com/example/marvelapp/presentation/MainActivity.kt [hunk 2]
```diff
@@ -20,21 +21,27 @@ class MainActivity : AppCompatActivity(), ItemClicked {
...
+                LoadCharactersUseCaseImpl(LocalMarvelDatabaseImpl(CharacterRealmMapper))
             ),
             MarvelAppView(this, binding)
         )
         presenter.fetchHeroes()
     }
 
+    private fun floatingButton() {
+        val button = binding.fab
+        button.setOnClickListener { presenter.getCharactersFromDataBase() }
```
Comment (mauroLonghi) [inline app/src/main/java/com/example/marvelapp/presentation/MainActivity.kt]: split in one line : binding.fab. setOnClickListener

File: app/src/main/java/com/example/marvelapp/presentation/MainActivity.kt [hunk 3]
```diff
@@ -20,21 +21,27 @@ class MainActivity : AppCompatActivity(), ItemClicked {
...
+                SaveCharactersDataBaseUseCaseImpl(LocalMarvelDatabaseImpl(CharacterRealmMapper)),
+                LoadCharactersUseCaseImpl(LocalMarvelDatabaseImpl(CharacterRealmMapper))
             ),
             MarvelAppView(this, binding)
         )
         presenter.fetchHeroes()
     }
 
+    private fun floatingButton() {
+        val button = binding.fab
```
Comment (LuchoElissondoGlobant) [inline app/src/main/java/com/example/marvelapp/presentation/MainActivity.kt]: no need of this variable.

File: app/src/main/java/com/example/marvelapp/presentation/mvp/presenter/MarvelAppPresenter.kt [hunk 4]
```diff
@@ -30,4 +30,13 @@ class MarvelAppPresenter(
...
 
+    override fun getCharactersFromDataBase() {
+        view.cleanRecycler()
+        view.showLoading()
+        model.loadCharactersFromDataBase().apply {
+            view.showData(this)
+            view.hideLoading()
+        }
+    }
+
```
Comment (LuchoElissondoGlobant) [inline app/src/main/java/com/example/marvelapp/presentation/mvp/presenter/MarvelAppPresenter.kt]: avoid unneccesary empty lines

File: app/src/main/java/com/example/marvelapp/data/mapper/CharacterRealmMapper.kt [hunk 5]
```diff
@@ -15,34 +15,82 @@ import com.example.marvelapp.domain.entity.StoriesEntity
...
+        transformDBListOfUrls(characterEntity.urls)
+    )
+
+    private fun transformCharacterEntityToRealm(characterRealmEntity: CharacterRealmEntity): CharacterEntity =
+        CharacterEntity(
+            characterRealmEntity.id,
+            characterRealmEntity.name,
+            characterRealmEntity.description,
+            characterRealmEntity.modified,
+            transformDBThumbnailr(characterRealmEntity.thumbnail!!),
```
Comment (LuchoElissondoGlobant) [inline app/src/main/java/com/example/marvelapp/data/mapper/CharacterRealmMapper.kt]: DO NOT USE null assert !!

File: app/build.gradle [hunk 6]
```diff
@@ -70,5 +69,6 @@ dependencies {
     implementation 'com.android.support:recyclerview-v7'
     // Glide
     implementation 'com.github.bumptech.glide:glide:4.11.0'
+    implementation 'com.google.android.material:material:1.0.0-rc01'
```
Comment (SartiMau) [inline app/build.gradle]: do not use rc versions pls

File: app/src/main/java/com/example/marvelapp/presentation/MainActivity.kt [hunk 7]
```diff
@@ -20,21 +21,27 @@ class MainActivity : AppCompatActivity(), ItemClicked {
...
-                SaveCharactersDataBaseUseCaseImpl(LocalMarvelDatabaseImpl(CharacterRealmMapper))
+                SaveCharactersDataBaseUseCaseImpl(LocalMarvelDatabaseImpl(CharacterRealmMapper)),
+                LoadCharactersUseCaseImpl(LocalMarvelDatabaseImpl(CharacterRealmMapper))
             ),
             MarvelAppView(this, binding)
         )
         presenter.fetchHeroes()
     }
 
+    private fun floatingButton() {
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/presentation/MainActivity.kt]: please put here a more representative name, this name doesn't make sense

File: app/src/main/java/com/example/marvelapp/data/local/LocalMarvelDatabaseImpl.kt [hunk 8]
```diff
@@ -9,12 +9,17 @@ import io.realm.Realm
 class LocalMarvelDatabaseImpl(private val mapper: CharacterRealmMapper) : LocalMarvelDataBase {
 
     override fun saveCharacters(data: List<CharacterEntity>) {
-        var realmInstance = Realm.getDefaultInstance()
-
+        val realmInstance = Realm.getDefaultInstance()
         realmInstance.executeTransaction {
             realmInstance.deleteAll()
-            var userRealm: List<CharacterRealmEntity> = mapper.transformDBListOfCharacters(data)
+            var userRealm: List<CharacterRealmEntity> = mapper.transformEntityListToRealmList(data)
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/data/local/LocalMarvelDatabaseImpl.kt]: This should be a val not var

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/marvelapp/data/mapper/CharacterRealmMapper.kt (hunk 5 above)
Comment (SartiMau) [reply app/src/main/java/com/example/marvelapp/data/mapper/CharacterRealmMapper.kt]: Do you know the difference between **!!** and **?.** ?

File: app/src/main/java/com/example/marvelapp/presentation/mvp/presenter/MarvelAppPresenter.kt (hunk 4 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/presentation/mvp/presenter/MarvelAppPresenter.kt]: Done thx

File: app/src/main/java/com/example/marvelapp/presentation/MainActivity.kt (hunk 7 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/presentation/MainActivity.kt]: Done thx

File: app/build.gradle (hunk 6 above)
Comment (Matias-Villalobo) [reply app/build.gradle]: Done thx

File: app/src/main/java/com/example/marvelapp/data/local/LocalMarvelDatabaseImpl.kt (hunk 8 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/data/local/LocalMarvelDatabaseImpl.kt]: Done thx

File: app/src/main/java/com/example/marvelapp/presentation/MainActivity.kt (hunk 3 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/presentation/MainActivity.kt]: Done thx

File: app/src/main/java/com/example/marvelapp/presentation/MainActivity.kt (hunk 2 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/presentation/MainActivity.kt]: Done thx

File: app/src/main/res/layout/activity_main.xml (hunk 1 above)
Comment (Matias-Villalobo) [reply app/src/main/res/layout/activity_main.xml]: Done thx

File: app/src/main/java/com/example/marvelapp/data/mapper/CharacterRealmMapper.kt (hunk 5 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/data/mapper/CharacterRealmMapper.kt]: The operator?, will verify if a variable is null before doing any operation with it, while the operator !!,  apply the operation without doing any verification

File: app/src/main/java/com/example/marvelapp/domain/entity/SeriesEntity.kt [hunk 9]
```diff
@@ -1,7 +1,10 @@
 package com.example.marvelapp.domain.entity
 
+import com.example.marvelapp.data.local.model.utils.ConstantsUtils.EMPTY_STRING
+import com.example.marvelapp.data.local.model.utils.ConstantsUtils.VALUE_ZERO
+
 class SeriesEntity(
-    val available: Int,
-    val collectionURI: String,
-    val returned: Int
+    val available: Int = VALUE_ZERO,
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/domain/entity/SeriesEntity.kt:7]: these default values VALUE_ZERO and EMPTY_STRING should be in the domain package. You should'n use the constants from data. Create a new ConstantsUtils here in domain

Comment (SartiMau) [reply app/src/main/java/com/example/marvelapp/domain/entity/SeriesEntity.kt:7]: check this for all the domain entities

**Review by LuchoElissondoGlobant — APPROVED** [review]

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/marvelapp/data/mapper/CharacterRealmMapper.kt (hunk 5 above)
Comment (SartiMau) [reply app/src/main/java/com/example/marvelapp/data/mapper/CharacterRealmMapper.kt]: great!

File: app/src/main/java/com/example/marvelapp/domain/entity/SeriesEntity.kt (hunk 9 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/domain/entity/SeriesEntity.kt:7]: Thx, now its done

**Review by mauroLonghi — APPROVED** [review]

**Review by SartiMau — APPROVED** [review]

**Pull request merged**